File: 030_Minimum Size Subarray Sum.py
"""
Minimum Size Subarray Sum

Find minimal length of contiguous subarray with sum >= target.
"""

from typing import List

def min_sub_array_len(target: int, nums: List[int]) -> int:
    n = len(nums)
    l = 0
    s = 0
    ans = float('inf')
    for r in range(n):
        s += nums[r]
        while s >= target:
            ans = min(ans, r - l + 1)
            s -= nums[l]
            l += 1
    return 0 if ans == float('inf') else ans


if __name__ == '__main__':
    print('MinSub 1:', min_sub_array_len(7, [2,3,1,2,4,3]))
    print('MinSub 2:', min_sub_array_len(4, [1,4,4]))
    print('MinSub 3:', min_sub_array_len(11, [1,2,3,4,5]))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 030_Minimum Size Subarray Sum.cpp
/*
Minimum Size Subarray Sum

Find minimal length of contiguous subarray with sum >= target.
*/

#include <bits/stdc++.h>
using namespace std;

int minSubArrayLen(int target, vector<int>& nums) {
    int n = nums.size();
    int l = 0, sum = 0, ans = INT_MAX;
    for (int r = 0; r < n; ++r) {
        sum += nums[r];
        while (sum >= target) {
            ans = min(ans, r - l + 1);
            sum -= nums[l++];
        }
    }
    return ans == INT_MAX ? 0 : ans;
}

int main() {
    vector<int> a1 = {2,3,1,2,4,3};
    cout << "MinSub 1: " << minSubArrayLen(7, a1) << endl; // 2

    vector<int> a2 = {1,4,4};
    cout << "MinSub 2: " << minSubArrayLen(4, a2) << endl; // 1

    vector<int> a3 = {1,2,3,4,5};
    cout << "MinSub 3: " << minSubArrayLen(11, a3) << endl; // 3
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
/*
Minimum Size Subarray Sum

Given an array of positive integers and a target, find the minimal length of a contiguous subarray of which the sum >= target. Return 0 if none.

Strategy: Sliding window two-pointer since numbers are positive.
*/

#include <vector>
#include <iostream>
#include <climits>
using namespace std;

int minSubArrayLen(int target, const vector<int>& nums) {
    int n = nums.size();
    int l = 0, sum = 0, ans = INT_MAX;
    for (int r = 0; r < n; ++r) {
        sum += nums[r];
        while (sum >= target) {
            ans = min(ans, r - l + 1);
            sum -= nums[l++];
        }
    }
    return ans == INT_MAX ? 0 : ans;
}

int main() {
    cout << "MinSub 1: " << minSubArrayLen(7, {2,3,1,2,4,3}) << '\n'; // 2
    cout << "MinSub 2: " << minSubArrayLen(4, {1,4,4}) << '\n'; // 1
    cout << "MinSub 3: " << minSubArrayLen(11, {1,2,3,4,5}) << '\n'; // 3
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/
